cocos2d/shaders/CCGLProgramState.js

```diff
@@ -0,0 +1,322 @@
+		this._type = types.GL_FLOAT_VEC2;
+	},
+
+	setVec3: function setVec3(v1, v2, v3) {
+		this._value = [v1, v2, v3];
+		this._type = types.GL_FLOAT_VEC3;
+	},
+
+	setVec3v: function setVec3v(value) {
+		this._value = value;
+		this._type = types.GL_FLOAT_VEC3;
+	},
+
+	setVec4: function setVec4(v1, v2, v3, v4) {
+		this._value = [v1, v2, v3, v4];
+		this._type = types.GL_FLOAT_VEC4;
+	},
+
+	setVec4v: function setVec4v(value) {
+		this._value = value;
+		this._type = types.GL_FLOAT_VEC4;
+	},
+
+	setMat4: function setMat4(value) {
+		this._value = value;
+		this._type = types.GL_FLOAT_MAT4;
+	},
+
+	setCallback: function setCallback(fn) {
+		this._value = fn;
+		this._type = types.GL_CALLBACK;
+	},
+
+	setTexture: function setTexture(textureId, textureUnit) {
+		this._value = textureUnit;
+		this._textureId = textureId;
+		this._type = types.GL_TEXTURE;
+	},
+
+	apply: function apply() {
+		switch (this._type) {
+			case types.GL_INT:
+				this._glprogram.setUniformLocationWith1i(this._uniform.location, this._value);
+				break;
+			case types.GL_FLOAT:
+				this._glprogram.setUniformLocationWith1f(this._uniform.location, this._value);
+				break;
+			case types.GL_FLOAT_VEC2:
+				this._glprogram.setUniformLocationWith2f(this._uniform.location, this._value[0], this._value[1]);
+				break;
+			case types.GL_FLOAT_VEC3:
+				this._glprogram.setUniformLocationWith3f(this._uniform.location, this._value[0], 
+					this._value[1], this._value[2]);
+				break;
+			case types.GL_FLOAT_VEC4:
+				this._glprogram.setUniformLocationWith4f(this._uniform.location, this._value[0], 
+					this._value[1], this._value[2], this._value[3]);
+				break;
+			case types.GL_FLOAT_MAT4:
+				this._glprogram.setUniformLocationWithMatrix4fv(this._uniform.location, this._value);
+				break;
+			case types.GL_CALLBACK:
+				this._value(this._glprogram, this._uniform);
+				break;
+			case types.GL_TEXTURE:
+				this._glprogram.setUniformLocationWith1i(this._uniform.location, this._value);
+				cc.glBindTexture2DN(this._value, this._textureId);
+				break;
+			default:
+				;
+		}
+	},
+});
+
+cc.GLProgramState = cc.Class.extend({
+	_glprogram: null,
+	_uniformsByName: null,
+	_uniformsByLocation: null,
+	_boundTextureUnits: null,
+	_textureUnitIndex: 0,
+
+	init: function init(glprogram) {
+		this._glprogram = glprogram;
+		this._uniformsByName = {};
+		this._uniformsByLocation = {};
+		this._boundTextureUnits = {};
+		this._textureUnitIndex = 1; // Start at 1, as CC_Texture0 is bound to 0
+		var userUniformKeys = Object.keys(glprogram._userUniforms);
+		for (var i = 0; i < userUniformKeys.length; ++i) {
+			var uniform = glprogram._userUniforms[userUniformKeys[i]];
+			var uniformValue = new cc.UniformValue(uniform, glprogram);
+			this._uniformsByName[uniform.name] = uniformValue;
+			this._uniformsByLocation[uniform.location] = uniformValue;
+		}
+	},
+
+	apply: function apply(modelView) {
+		this.applyGLProgram(modelView);
+		this.applyUniforms();
```

Comment:
We can check for modelView before invoking applyGLProgram, we need to reduce as many js function calls as possible.

And we don't need to apply uniform in each rendering call, do we ?